File: Programmers/level2/changeBracket/change_bracket.py
# 문제 출처 : https://programmers.co.kr/learn/courses/30/lessons/60058
# 참고 블로그 : https://johnyejin.tistory.com/124

def solution(p):
    answer = ''
    stack = []
    print(check(p))
    print(divide(p))
    while len(p) != 0:
        u, p = divide(p)
        if check(u):
            answer += u
        else:
            answer += '(' + solution(p) + ')' + reverse(u[1:-1])
            break
    print(answer)
    return answer


def check(p):
    stack = []
    for i in p:
        if i == "(":
            stack.append("(")
        elif i == ")":
            if not stack:
                return False
            else:
                stack.pop()
    return True


def divide(p):
    cnt = [0, 0]
    for i in p:
        if i == "(":
            cnt[0] += 1
        else:
            cnt[1] += 1
        if cnt[0] == cnt[1]:
            break
    return p[:sum(cnt)], p[sum(cnt):]


def reverse(v):
    temp = ''
    for i in v:
        if i == '(':
            temp += ')'
        else:
            temp += '('
    return temp

File: Programmers/level2/changeBracket/changeBracket.js
export const solution = (p) => {
    var answer = '';
    
    while(p.length !== 0){
        let u = divide(p)[0];
        p = divide(p)[1] ;
        // console.log('u : ', u);
        // console.log('p : ', p)
        if(check(u)){
            answer += u
        } else {
            answer += '(' + solution(p) + ')' + reverse(u.slice(1,-1))
            break
        }
    }
    return answer;
}

const divide = (p) => {
    let cnt = [0, 0]
    for(let i = 0; i < p.length; i++){
        if(p[i] === '('){
            cnt[0] += 1;
        }else{
            cnt[1] += 1;
        }
        if(cnt[0] === cnt[1]){
            break;
        }
    }
    let sum = cnt[0] + cnt[1];
    let result = [p.slice(0,sum), p.slice(sum,)]
    return result
}

const check = (s) => {
    let stack = [];
    for(let i = 0; i < s.length; i++){
        if(s[i] === '('){
            stack.push(')');
        }
        if(s[i] === ')'){
            if(stack.length === 0){
                return false
            }
            stack.pop()
        }
    return true;
    }
}

const reverse = (v) => {
    let temp = '';
    for(let i = 0; i < v.length; i++){
        if(v[i] === '('){
            temp += ')';
        } else {
            temp += '(';
        }
    }
    return temp;
}